source/resource_name.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/resource_name.h>
+
+#define ARN_DELIMETER ":"
+
+static int s_parse_resource_name_part(
+    struct aws_byte_cursor *token_cursor,
+    const struct aws_byte_cursor *cursor_split_on,
+    struct aws_byte_cursor *prev_token_cursor) {
+    prev_token_cursor->ptr = token_cursor->ptr + 1;
+    prev_token_cursor->len = token_cursor->len;
+
+    if (aws_byte_cursor_find_exact(prev_token_cursor, cursor_split_on, token_cursor)) {
+        return aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+    }
+    prev_token_cursor->len = token_cursor->ptr - prev_token_cursor->ptr;
+    return AWS_OP_SUCCESS;
+}
+
+AWS_COMMON_API
+int aws_resource_name_init_from_cur(struct aws_resource_name *arn, const struct aws_byte_cursor *input) {
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(input));
+    AWS_PRECONDITION(input->ptr);
+    AWS_PRECONDITION(arn);
+
+    const struct aws_byte_cursor cursor_split_on = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(ARN_DELIMETER);
```

Comment:
This whole thing seems unnecessarily complicated. We have a split api. Why not?

1. split on the delimiter
2. Verify the split count == 5
3. Verify first split is "arn"
4. Assign each split segment to the appropriate resource name part
5. done

Reply:
I was avoiding the edge case around the there not always being 5 parts to an arn and my initial understanding of aws_array_list was that it would trigger unnecessary allocations. Once the resource ID part is found at the end, the rest of the characters are in it.

Reply:
Array list has a static mode. That will not perform any allocations.

Reply:
I went with not doing any allocations, in favor of the caller taking care of it however they see fit and failing if they have not provided enough space.

Read the impl o aws_byte_cursor_split_on_char_n() and it's doing what I need. Going to change it up one more time.